src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1392,21 +1389,40 @@ private bool HasIdentityConversionInternal(TypeSymbol type1, TypeSymbol type2)
             return HasIdentityConversionInternal(type1, type2, IncludeNullability);
         }
 
-        internal static bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
+        internal bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
         {
-            bool? sourceIsNullable = source.IsNullable;
-            bool? destinationIsNullable = destination.IsNullable;
-            return sourceIsNullable == null || destinationIsNullable == null || sourceIsNullable == destinationIsNullable;
-        }
-
-        internal static bool HasTopLevelNullabilityImplicitConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
-        {
-            return HasTopLevelNullabilityImplicitConversion(source.IsNullable, destination.IsNullable);
+            if (!IncludeNullability)
+            {
+                return true;
+            }
+            return source.IsNullable == destination.IsNullable;
         }
 
-        internal static bool HasTopLevelNullabilityImplicitConversion(bool? sourceIsNullable, bool? destinationIsNullable)
+        internal bool HasTopLevelNullabilityImplicitConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
         {
-            return sourceIsNullable != true || destinationIsNullable != false;
+            if (!IncludeNullability)
+            {
+                return true;
+            }
+            bool? sourceIsNullable = source.IsNullable;
+            bool? destinationIsNullable = destination.IsNullable;
+            if (sourceIsNullable == destinationIsNullable)
+            {
+                return true;
+            }
+            if (sourceIsNullable == null || destinationIsNullable == null)
+            {
+                return true;
+            }
+            switch (sourceIsNullable)
+            {
+                case true:
+                    return false;
+                case false:
+                    return true;
+                default:
+                    return destinationIsNullable == true;
```

Comment:
Considering that we've already checked sourceIsNullable for null, is this codepath hittable?

Reply:
Feels like it would be. One test regards the source, but the other concerns the destination.

---
In reply to: [223483964](https://github.com/dotnet/roslyn/pull/29727#discussion_r223483964) [](ancestors = 223483964)